### ISSUE TITLE
generic type instance contains RecWhen, causing internal errors during code generation

### ISSUE THREAD

**Issue by mratsim** [nim-lang/Nim#9678]
Maybe related to #7378, another issue with `when` in type section, this time with generics.

Error is `Error: illformed AST: result.my_enum` when assigning
or `internal error: (filename: "semobjconstr.nim", line: 242, column: 19)` if constructing the object.

Solving the error requires either not using generics or not using `when`

```Nim
type
  Foo*[T] = object                        # Remove generic to solve the issue
    bar*: int
    when defined(i386) or defined(amd64): # or Remove this to solve the issue
      my_enum: MyEnum

  MyEnum* = enum
    choice1,
    choice2

func fails_IllformedAST*(
      T: type,
      choice: MyEnum
    ): Foo[T] =
  result.my_enum = choice

func fails_ICE*(
      T: type,
      choice: MyEnum
    ): Foo[T] =
  Foo[T](bar: 1, my_enum: choice)

# let test1 = fails_IllformedAST(int, choice1) # Error: illformed AST: result.my_enum
let test2 = fails_ICE(int, choice1)            # Error: internal error: (filename: "semobjconstr.nim", line: 242, column: 19)


############################################
# The following works

# type
#   FooNoGeneric* = object                  # Remove generic to solve the issue
#     bar*: int
#     when defined(i386) or defined(amd64): # or Remove this to solve the issue
#       my_enum*: MyEnum

#   MyEnum* = enum
#     choice1,
#     choice2

# func works1*(
#       T: type,
#       choice: MyEnum
#     ): FooNoGeneric =
#   result.my_enum = choice

# func works2*(
#       T: type,
#       choice: MyEnum
#     ): FooNoGeneric =
#   FooNoGeneric(bar: 1, my_enum: choice)

# let test1_nogen = works1(int, choice1)
# let test2_nogen = works2(int, choice1)
```

**Comment by mratsim** [comment]
Dropping to low priority, we can "solve" the issue with a dummy field:

```Nim
type
  Foo*[T] = object                        # Remove generic to solve the issue
    bar*: int
    dummy: T
    when defined(i386) or defined(amd64): # or Remove this to solve the issue
      my_enum: MyEnum
```

**Comment by krux02** [comment]
```nim
type
  Foo*[T] = object                        # Remove generic to solve the issue
    bar*: int
    when defined(i386) or defined(amd64): # or Remove this to solve the issue
      my_enum: MyEnum

  MyEnum* = enum
    choice1,
    choice2

import macros

macro foobar(arg: typed): untyped =
  echo arg.getTypeImpl.treeRepr

var x : Foo[int]

foobar(x)
```

This still crashes during code generation pass. Aparently the when statement is not unrolled properly during gerenic expansion. This is the internal type AST.

```
ObjectTy
  Empty
  Empty
  RecList
    IdentDefs
      Sym "bar"
      Sym "int"
      Empty
    RecWhen
      ElifBranch
        IntLit 1
        IdentDefs
          Sym "my_enum"
          Sym "MyEnum"
          Empty
```

A generic instance should never contain a when statement.

**Comment by mratsim** [comment]
Your last sentence is a bit ambiuguous:
Do you mean "Once instantiated, the RecWhen should be 'resolved' and not appear in the AST" or do you mean "Generic types should not have `when` in a type section".

Note that when the fields exist in both when branches and are just merely reordered, there is no issue. The following type as been working daily since March: https://github.com/status-im/nim-stint/blob/edb1ade37309390cc641cee07ab62e5459d9ca44/stint/private/datatypes.nim#L141-L164

```Nim
type
  # ### Private ### #
  UintImpl*[BaseUint] = object
    when system.cpuEndian == littleEndian:
      lo*, hi*: BaseUint
    else:
      hi*, lo*: BaseUint

  # ### Private ### #

  StUint*[bits: static[int]] = object
    data*: uintImpl(bits)
```

**Comment by krux02** [comment]
I mean "Once instantiated, the RecWhen should be 'resolved' and not appear in the AST". It can and must be resolved during template instantiation. Any RecWhen that can't be resolved during template instantiation must be a compile error, because the correct set of members can't be figured out. Just think of it that a RecWhen would be allowed, every macro that accesses the type ast would be forced to special case for the RecWhen, almost every macro would do it initially wrong, because the developers won't be aware that this is a thing. And I can't imagine any benefit for it.

**Comment by LemonBoy** [comment]
This has been fixed by #8748! :tada:
